driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2262,12 +2262,12 @@ private Connection prepareAllQueries(Host host, Connection reusedConnection) thr
                 // used for preparing it. However, since we are likely that all prepared query belong to only a handful
                 // of different keyspace (possibly only one), and to avoid setting the current keyspace more than needed,
                 // we first sort the query per keyspace.
-                SetMultimap<String, String> perKeyspace = HashMultimap.create();
+                SetMultimap<String, PreparedStatement> perKeyspace = HashMultimap.create();
```

Comment:
Could also change this such that if a PreparedStatement has a keyspace set explicitly, prepare it with default session since it doesn't matter what the session keyspace state is in this case.